numpy/fft/_pocketfft_umath.c

```diff
@@ -0,0 +1,354 @@
+        /* TODO: Requires use of new ufunc API to indicate error return */
+        NPY_ALLOW_C_API_DEF
+        NPY_ALLOW_C_API;
+        PyErr_NoMemory();
+        NPY_DISABLE_C_API;
+    }
+    return;
+}
+
+/*
+ * For the forward real, we cannot know what the requested number of points is
+ * just based on the number of points in the complex output array (e.g., 10
+ * and 11 real input points both lead to 6 complex output points), so we
+ * define versions for both even and odd number of points.
+ */
+static void
+rfft_n_even_loop(char **args, npy_intp const *dimensions, npy_intp const *steps, void *func)
+{
+    npy_intp npts = 2 * dimensions[2] - 2;
+    rfft_impl(args, dimensions, steps, func, npts);
+}
+
+static void
+rfft_n_odd_loop(char **args, npy_intp const *dimensions, npy_intp const *steps, void *func)
+{
+    npy_intp npts = 2 * dimensions[2] - 1;
+    rfft_impl(args, dimensions, steps, func, npts);
+}
+
+
+static void
+irfft_loop(char **args, npy_intp const *dimensions, npy_intp const *steps, void *func)
+{
+    char *ip = args[0], *fp = args[1], *op = args[2];
+    npy_intp n_outer = dimensions[0];
+    npy_intp si = steps[0], sf = steps[1], so = steps[2];
+    npy_intp nin = dimensions[1], nout = dimensions[2];
+    npy_intp isi = steps[3], iso = steps[4];
+    npy_intp npts = nout;
+    char *buffer = NULL;
+    int no_mem = 1;
+
+    rfft_plan plan = make_rfft_plan(npts);
+    if (plan == NULL) {
+        goto fail;
+    }
+    if (iso != sizeof(npy_double)) {
+        buffer = malloc(npts * sizeof(npy_double));
+        if (buffer == NULL) {
+            goto fail;
+        }
+    }
+
+    for (npy_intp i = 0; i < n_outer; i++, ip += si, fp += sf, op += so) {
+        double fct = *(double *)fp;
+        char *op_buf = buffer == NULL? op : buffer;
+        double *op_double = (double *)op_buf;
+        /* Copy complex input to buffer in FFTpack order */
+        op_double[0] = ((double *)ip)[0];
+        if (nin > 1) {
+            copy_data(ip + isi, isi, nin - 2,
+                      (char *)&op_double[1], sizeof(npy_cdouble), npts / 2 - 1,
+                      sizeof(npy_cdouble));
+            npy_intp ncopied = (npts / 2 - 1  < nin - 2 ? npts / 2 - 1 : nin - 2);
+            double *last = (double *)(ip + (ncopied + 1) * isi);
+            op_double[ncopied * 2 + 1] = last[0];
+            if (npts % 2 == 1) {  /* For odd n, we still imag real of the last point */
+                op_double[ncopied * 2 + 2] = last[1];
+            }
+        }
+        if ((no_mem = rfft_backward(plan, op_double, fct)) != 0) {
+            break;
+        }
+        if (op_buf != op) {
+            copy_data(op_buf, sizeof(npy_double), npts, op, iso, npts, sizeof(npy_double));
+        }
+    }
+  fail:
+    if (buffer != NULL) {
+        free(buffer);
+    }
+    if (plan != NULL) {
+        destroy_rfft_plan(plan);
+    }
+    if (no_mem) {
+        /* TODO: Requires use of new ufunc API to indicate error return */
```

Comment:
Also, the new API may be able to move this to before the loop is even called or at least avoid repeating it (not that it is usually much of a problem).

Reply:
Yes, this did seem a good case for trying the new API. I thought about it, together with the C++ templating, but I did not quite know how to do some of the more specific non-type differences between the routines (e.g., the `inplace` only being relevant for complex<->complex). Maybe I'll still try...